feat(store,world): replace `ResourceSelector` with `ResourceId` and `WorldResourceId` (#1544)
latticexyz/mud: .changeset/stale-seahorses-pay.md

---
"@latticexyz/cli": major
"@latticexyz/common": major
"@latticexyz/config": major
"@latticexyz/store": major
---

- `ResourceSelector` is replaced with `ResourceId`, `ResourceIdLib`, `ResourceIdInstance`, `WorldResourceIdLib` and `WorldResourceIdInstance`.

  Previously a "resource selector" was a `bytes32` value with the first 16 bytes reserved for the resource's namespace, and the last 16 bytes reserved for the resource's name.

>   Now a "resource ID" is a `bytes32` value with the first 14 bytes reserved for the resource's namespace, the next 16 bytes reserved for the resource's name, and the last 2 bytes reserved for the resource type.

Author: ```suggestion
  Now a "resource ID" is a `bytes32` value with the first 2 bytes reserved for the resource type, the next 14 bytes reserved for the resource's namespace, and the last  16 bytes reserved for the resource's name.
```

  Previously `ResouceSelector` was a library and the resource selector type was a plain `bytes32`.
  Now `ResourceId` is a user type, and the functionality is implemented in the `ResourceIdInstance` (for type) and `WorldResourceIdInstance` (for namespace and name) libraries.

>   We split the logic into two libraries, because `Store` also uses `ResourceId` now and needs awareness of resource types, but not of namespaces/names.

Author: ```suggestion
  We split the logic into two libraries, because `Store` now also uses `ResourceId` and needs to be aware of resource types, but not of namespaces/names.
```

  ```diff
  - import { ResourceSelector } from "@latticexyz/world/src/ResourceSelector.sol";
  + import { ResourceId, ResourceIdInstance } from "@latticexyz/store/src/ResourceId.sol";
  + import { WorldResourceIdLib, WorldResourceIdInstance } from "@latticexyz/world/src/WorldResourceId.sol";
  + import { RESOURCE_SYSTEM } from "@latticexyz/world/src/worldResourceTypes.sol";

  - bytes32 systemId = ResourceSelector.from("namespace", "name");

>   + ResourceId systemId = WorldResourceIdLib.encode("namespace", "name", RESOURCE_SYSTEM);

Author: ```suggestion
  + ResourceId systemId = WorldResourceIdLib.encode(RESOURCE_SYSTEM, "namespace", "name");
```

  - using ResourceSelector for bytes32;
  + using WorldResourceIdInstance for ResourceId;
  + using ResourceIdInstance for ResourceId;

    systemId.getName();
    systemId.getNamespace();
  + systemId.getType();

  ```

- All `Store` and `World` methods now use the `ResourceId` type for `tableId`, `systemId`, `moduleId` and `namespaceId`.
  All mentions of `resourceSelector` were renamed to `resourceId` or the more specific type (e.g. `tableId`, `systemId`)

>   ```diff
>   import { ResourceId } from "@latticexyz/store/src/ResourceId.sol";
> 
>   IStore {
>     function setRecord(
>   -   bytes32 tableId,
>   +   ResourceId tableId,
>       bytes32[] calldata keyTuple,
>       bytes calldata staticData,
>       PackedCounter encodedLengths,
>       bytes calldata dynamicData,
>       FieldLayout fieldLayout
>     ) external;
> 
>     // Same for all other methods
>   }
>   ```

Reviewer: ```suggestion
  ```diff
    import { ResourceId } from "@latticexyz/store/src/ResourceId.sol";
  
    IStore {
      function setRecord(
  -     bytes32 tableId,
  +     ResourceId tableId,
        bytes32[] calldata keyTuple,
        bytes calldata staticData,
        PackedCounter encodedLengths,
        bytes calldata dynamicData,
        FieldLayout fieldLayout
      ) external;
  
      // Same for all other methods
    }
  ```
```

  ```diff
  import { ResourceId } from "@latticexyz/store/src/ResourceId.sol";

>   IBaseWorld {
>   function callFrom(
>     address delegator,
>   -   bytes32 resourceSelector,
>   +   ResourceId systemId,
>     bytes memory callData
>   ) external payable returns (bytes memory);
> 
>   // Same for all other methods
>   }

Author: ```suggestion
  IBaseWorld {
    function callFrom(
      address delegator,
  -   bytes32 resourceSelector,
  +   ResourceId systemId,
      bytes memory callData
    ) external payable returns (bytes memory);

    // Same for all other methods
  }
```

>   ```

Reviewer: ugh gihub messed up my attempt at suggesting a chance, but I think for diffs you want a "gutter" where the + and - live 

so basically everything in the code block should indent one level then dedent the +- to be at the left edge
(see your diff above)

>   ```

Author: oof yes that's what i did, i think prettier keeps messing this up again
